uamqp/client.py

```diff
@@ -640,13 +640,24 @@ def receive_message_batch(self, max_batch_size=None, on_message_received=None, t
             return batch
 
         while receiving and not expired and len(batch) < max_batch_size:
-            receiving = self.do_work(batch=max_batch_size)
+
+            while receiving and self._received_messages.qsize() < max_batch_size:
+                if timeout and time.time() > timeout:
+                    expired = True
+                    break
+
+                before = self._received_messages.qsize()
+                receiving = self.do_work(batch=max_batch_size)
+                received = self._received_messages.qsize() - before
+
+                if self._received_messages.qsize() > 0 and received == 0:
+                    expired = True
```

Comment:
Why would this condition result in expiry?

Reply:
the naming `expired` is actually a bit ambiguous here, I have removed this variable to just use flag `receiving`.